File: algorithms/714. Best Time to Buy and Sell Stock with Transaction Fee/solution.cc
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        // @zestypanda
        int s0 = 0, s1 = INT_MIN, tmp;
        for ( int p : prices ) {
            tmp = s0;
            s0 = max(s0, s1 + p);
            s1 = max(s1, tmp - p - fee);
        }
        return s0;
    }
};

// 44 / 44 test cases passed.
// Status: Accepted
// Runtime: 139 ms

File: algorithms/714. Best Time to Buy and Sell Stock with Transaction Fee/solution.py
class Solution(object):
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        # inspired by @zestypanda 
        s0, s1 = 0, float('-inf')
        for p in prices:
            s0, s1 = max(s0, s1 + p - fee), max(s1, s0 - p)
        return s0

# 44 / 44 test cases passed.
# Status: Accepted
# Runtime: 335 ms
